[AlyssaPng] iP

AlyssaPng opened the pull request:
### Chatty - Your very own chatbot!!

> “The shorter way to do many things is to only do one thing at a time.” — Mozart [(source)](https://deepstash.com/idea/18151/the-shorter-way-to-do-many-things-is-to-only)

Chatty helps with task management and frees your mind of things you need to complete. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/AlyssaPng/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [X] Managing tasks (Todo, Deadlines and Events)
- [ ]  Reminders (coming soon)
- [ ] Arrange by dates (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method in `Launcher` :
 
```ruby
  public class Launcher {
      public static void main(String[] args) {
          Application.launch(Main.class, args);
      }
  }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

AlyssaPng committed:
Level-0: Changed to Chatty

AlyssaPng committed:
Level-1: Added in userinput

AlyssaPng committed:
Level-2: Added List

AlyssaPng committed:
Level-3: Changed to Arraylist and added Task class

AlyssaPng committed:
Level-3: Added Todo, Deadline, Even classes and change Duke.java

AlyssaPng committed:
Automated Text UI Testing

AlyssaPng committed:
Level-5: Handle Errors by adding DukeException and subexceptions

AlyssaPng committed:
Level-6 - adding delete option and javadocs

AlyssaPng committed:
Add methods to store file data

AlyssaPng committed:
Merge branch 'branch-Level-7'

AlyssaPng committed:
Add in DateTime Formatter

AlyssaPng committed:
Merge branch 'branch-Level-8'

AlyssaPng committed:
Add more classes for more OOP

AlyssaPng committed:
Merge branch 'branch-A-MoreOOP'

AlyssaPng committed:
Add duke and exception packages

AlyssaPng committed:
Merge remote-tracking branch 'origin/add-gradle-support'

AlyssaPng committed:
Add JUnit Test files and cases

AlyssaPng committed:
Add JUnit Test files and cases

AlyssaPng committed:
Add JAR dependencies

AlyssaPng committed:
Merge branch 'master' of https://github.com/AlyssaPng/ip

AlyssaPng committed:
Add JavaDoc to JUnit test classes

AlyssaPng committed:
Update code to be comply with coding standard

AlyssaPng committed:
Add find command

AlyssaPng committed:
Merge branch 'branch-A-JavaDoc'

AlyssaPng committed:
Merge branch 'branch-A-CodingStandard'

AlyssaPng committed:
Merge branch 'branch-Level-9'

geoffong11 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,113 @@
+     */
+    public Task loadData(String dataInput) throws DukeException {
+        Task newTask;
+        String[] arr = dataInput.split("\\|");
```
Would dataInputArray be a better name for the parameter?

geoffong11 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,234 @@
+     */
+    public Task createNewTask(String userInput) throws DukeException {
+        Task newTask;
+        String[] arr = userInput.split(" ", 2);
```
Would userInputArray be a better name for the parameter?

geoffong11 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,61 @@
+     */
+    @Override
+    public String toString() {
+        return ("[" + getStatusIcon() + "] " + this.description);
```
I like how the status icon is abstracted out!

geoffong11 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,130 @@
+        } else if (isFindCommand(userInput)) {
+            String keyword = userInput.replaceFirst("find", "").trim();
+            if (keyword.isEmpty()) {
+                throw new InvalidInputException("☹ OOPS!!! I'm sorry, please provide a keyword to search.");
```
I like how the throwing of exceptions is specialized for every exception!

geoffong11 reviewed (commented):
Looks good 👍

AlyssaPng committed:
Add in checkstyle files and update code accordingly

AlyssaPng committed:
Add GUi for Chatty

shishirbychapur commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,79 @@
+     * @throws DukeException If an error has occurred in the duke.Duke application.
+     */
+    public static void main(String[] args) {
+        new Duke("C:/repos/ip/src/main/java/data/duke.txt").run();
```
Not a good idea to use absolute path, works for now, but may not work on client's side!

shishirbychapur commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    /**
+     * Overides toString method from duke.Task
```
Small nit (Misspelt Override)

shishirbychapur commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    /**
+     * Overides toString method from duke.Task
```
Misspelt Override again (Small nit)

shishirbychapur commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,79 @@
+                ui.showMessage(e.getMessage());
+            }
+        }
+    }
```
Can consider removing this since Duke uses GUI.

shishirbychapur commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,79 @@
+    public Duke() {
+        ui = new Ui();
+        tasks = new TaskList();
+    }
```
If no file path is provided, Storage isn't initialized? Can consider adding a default file path like "data/duke.txt".

shishirbychapur commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,19 @@
+        Application.launch(Main.class, args);
+    }
+}
+
```
Can consider deleting this launcher class, and launching the GUI directly from the Duke class.

shishirbychapur commented on src/main/java/duke/exception/NoEndDateException.java:
Don't really see the purpose of creating exception classes which extend Duke Exception (There could be a lot of edge cases, and it isn't viable to create one for every case). Instead, you can just throw a new Duke Exception with the required message and action.

For example:
```java
if (!event.equals("todo")) {
    throw new DukeException("Invalid todo event");
}
```

shishirbychapur commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,76 @@
+ */
+public class Ui {
+
+    private Scanner scanner;
```
Can consider removing this, since your program makes use of a GUI (unless you want to continue using CLI)

shishirbychapur commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,112 @@
+                throw new EmptyDateException("deadline");
+            }
+
+            newTask = new Deadline(arr[2], arr[3]);
```
Better to use switch case here? Bit hard to read your code here since there are a lot of if and else statements.

shishirbychapur reviewed (commented):
Overall, made use of many abstractions and many exception classes (some of which can be unnecessary), and no blatant code violations. Made effort to beautify GUI (inferred from the CSS in the DialogBox and MainWindow Classes). Good work!

AlyssaPng committed:
Update code using comments from peer review

AlyssaPng committed:
Add assertion commands

There are several sections of the code where assumptions about the program state are made.

Adding these assertions can ensure that when the execution comes to the assert command, these assumptions are verified by the runtime.

Let's add assertions in
* Duke class to ensure the TaskList is intialised before running application.
* Mainwindow class to make sure that a String is return from a getresponse() before creating Diaglog object.
* Parser class to ensure that the userInput parsed in is not empty and if-statement elimates all cases where the keyword is empty.
* Tasklist class to ensure the tasks is successfully added, marked and unmarked or that else statement handles Event tasks.

assertions ensure the assumptions made are correct and that no mistake in the code goes unnoticed.

AlyssaPng committed:
Update code to improve code quality.

Code includes long methods, recycled variables and other code quality issues. With these, the code is hard to read and "unsafe".

Updating the code accordingly will imrpove the code

Lets'
* Rename recycled variables
* Add helper methods to shorten long methods
* Change the code to focus on the happy path

AlyssaPng committed:
Merge pull request #1 from AlyssaPng/branch-A-Assertions

Add assertion commands

AlyssaPng committed:
Merge branch 'master' into branch-A-CodeQuality

AlyssaPng committed:
Merge branch 'master' of https://github.com/AlyssaPng/ip

AlyssaPng committed:
Add the ability to recognize and deal with duplicate items

The user can to add duplicate task in the task list making which compromising the effectiveness of Chatty, the task management system.

Adding the ability to recognise and deal with duplicate items will prevent the list from having 2 or more of the same task.

Lets change the parent class Task to abstract and add an abstract method to check if the new Task Object is a duplicate of the curretn Task. Event, ToDo and Deadline class override this abstract method and check the different fields accordingly.

Add the ability to recognize and deal with duplicate items

The user can to add duplicate task in the task list making which compromising the effectiveness of Chatty, the task management system.

Adding the ability to recognise and deal with duplicate items will prevent the list from having 2 or more of the same task.

Lets change the parent class Task to abstract and add an abstract method to check if the new Task Object is a duplicate of the curretn Task. Event, ToDo and Deadline class override this abstract method and check the different fields accordingly.

This follows the OOP properties as it assigns the responsibility of duplicate checking to each subclass, ensuring that they manage duplicates within their respective domains.

AlyssaPng committed:
Merge branch 'branch-C-DetectDuplicates'

AlyssaPng committed:
Update user guide

AlyssaPng committed:
Update code for application release

AlyssaPng committed:
Update ReadME and code to imporve application

AlyssaPng committed:
Update code to finailise ip

AlyssaPng committed:
Update code to improve code standards

AlyssaPng committed:
Merge branch 'master' into branch-A-CodeQuality

AlyssaPng committed:
Merge pull request #2 from AlyssaPng/branch-A-CodeQuality

Update code to improve code quality.

AlyssaPng committed:
Update Parser.java

AlyssaPng committed:
Merge branch 'master' of https://github.com/AlyssaPng/ip

AlyssaPng committed:
Change code to integrate merge

AlyssaPng committed:
Update ReadME and GUI user interface.

AlyssaPng committed:
Change spelling errors in ReadME

AlyssaPng committed:
Update code to handle errors.

AlyssaPng committed:
Update Ui.png

AlyssaPng committed:
Change code to standardise input

AlyssaPng committed:
Update README

AlyssaPng committed:
Update GUI and code quality.

AlyssaPng committed:
Merge branch 'branch-Level-10'

AlyssaPng committed:
Change DialogBox.ftml to fix warnings.

AlyssaPng committed:
Change Error Handling

AlyssaPng committed:
Update JavaDocs

AlyssaPng committed:
Update TaskList

AlyssaPng committed:
Delete Unneccessary File

AlyssaPng committed:
Updated code to remove white spaces.

AlyssaPng committed:
Update build.grader to fix bugs
